datasets/flwr_datasets/federated_dataset.py

```diff
@@ -0,0 +1,144 @@
+
+     Partitions are created using IidPartitioner. Support for different partitioners
+     specification and types will come in future releases.
+
+    Parameters
+    ----------
+    dataset: str
+        The name of the dataset in the HuggingFace Hub.
+    partitioners: Dict[str, int]
+        Dataset split to the number of iid partitions.
+
+    Examples
+    --------
+    Use MNIST dataset for Federated Learning with 100 clients (edge devices):
+
+    >>> mnist_fds = FederatedDataset(dataset="mnist", partitioners={"train": 100})
+
+    Load partition for client with id 10.
+
+    >>> partition = mnist_fds.load_partition(10, "train")
+
+    Use test split for centralized evaluation.
+
+    >>> centralized = mnist_fds.load_full("test")
+    """
+
+    def __init__(self, *, dataset: str, partitioners: Dict[str, int]) -> None:
+        _check_if_dataset_supported(dataset)
+        self._dataset_name: str = dataset
+        self._partitioners: Dict[str, Partitioner] = _instantiate_partitioners(
+            partitioners
+        )
+        #  Init (download) lazily on the first call to `load_partition` or `load_full`
+        self._dataset: Optional[DatasetDict] = None
+
+    def load_partition(self, idx: int, split: str) -> Dataset:
+        """Load the partition specified by the idx in the selected split.
+
+        Parameters
+        ----------
+        idx: int
+            Partition index for the selected split, idx in {0, ..., num_partitions - 1}.
+        split: str
+            Split name of the dataset to partition (e.g. "train", "test").
+
+        Returns
+        -------
+        partition: Dataset
+            Single partition from the dataset split.
+        """
+        self._download_dataset_if_none()
+        if self._dataset is None:
+            raise ValueError("Dataset is not loaded yet.")
+        self._check_if_split_present(split)
+        self._check_if_split_possible_to_federate(split)
+        partitioner: Partitioner = self._partitioners[split]
+        self._assign_dataset_to_partitioner(split)
+        return partitioner.load_partition(idx)
+
+    def load_full(self, split: str) -> Dataset:
+        """Load the full split of the dataset.
+
+        Parameters
+        ----------
+        split: str
+            Split name of the downloaded dataset (e.g. "train", "test").
+
+        Returns
+        -------
+        dataset_split: Dataset
+            Part of the dataset identified by its split name.
+        """
+        self._download_dataset_if_none()
+        if self._dataset is None:
+            raise ValueError("Dataset is not loaded yet.")
+        self._check_if_split_present(split)
+        return self._dataset[split]
+
+    def _download_dataset_if_none(self) -> None:
+        """Download dataset if the dataset is None - meaning it is not downloaded yet.
+
+        The dataset is downloaded only when the first call to `load_partition` or
+        `load_full` is made.
```

Comment:
This comment should be part of the calling code (this function can not decide whether it is being called or not)

Reply:
It's now added to both the `load_full` and `load_partition` (the only two functions that can trigger that and use the `_download_dataset_if_none` method.